onnxruntime/core/session/inference_session.h

```diff
@@ -867,14 +866,18 @@ class InferenceSession {
       return cached_execution_provider_for_graph_replay_ != nullptr && cached_execution_provider_for_graph_replay_->IsGraphCaptureEnabled();
     }
 
-    bool IsGraphCaptured() const {
-      return cached_execution_provider_for_graph_replay_ != nullptr && cached_execution_provider_for_graph_replay_->IsGraphCaptured();
+    bool IsGraphCaptured(int graph_annotation_id) const {
+      return cached_execution_provider_for_graph_replay_ != nullptr && cached_execution_provider_for_graph_replay_->IsGraphCaptured(graph_annotation_id);
+    }
+
+    bool AllowGraphCaptureOnRun(int graph_annotation_id) const {
+      return cached_execution_provider_for_graph_replay_ != nullptr && graph_annotation_id != kGraphAnnotationSkip;
     }
 
-    Status ReplayGraph() {
-      ORT_ENFORCE(IsGraphCaptured());
+    Status ReplayGraph(int graph_annotation_id) {
+      ORT_ENFORCE(IsGraphCaptured(graph_annotation_id));
```

Comment:
Nit: Maybe we can clean up superfluous sanity checks. `IsGraphCaptured()` seems superfluous as the necessary condition to trigger `ReplayGraph()` is for `IsGraphCaptured()` to be true (see logic in inference_session.cc). In any case, we will do a similar sanity check while checking if the annotation_id exists in the map in `CUDAGraphManager`.